### ISSUE TITLE
concatAll

### ISSUE THREAD

**Issue by joaomilho** [ramda/ramda#1897]
I've read all previous attempts to have a variadic concat in the lib, and I understand the reasoning for denying. Nevertheless, there's a merge and a mergeAll, which do pretty much the same for objects. It seems to me that for consistency sake a concatAll that accepts 1 array would make sense. WDYT?

**Comment by buzzdecafe** [comment]
i would need more convincing. If you want this function it is trivial to implement as `const concatAll = reduce(concat, []);` which doesn't seem too onerous to me.

**Comment by joaomilho** [comment]
That's what I did :D (https://github.com/ion-lang/ion/commit/e61baa01d53a31e6c49981a59289da7dcf1c1812#diff-a955ca36ac676ea98972190d6bf81299R186) But a mergeAll is also simple to implement with more basic functions, isn't it?

**Comment by davidchambers** [comment]
If we're okay with `R.mergeAll` we should be okay with `R.concatAll`. Interestingly, if `R.concat` merged string maps (which it should) we wouldn't need `R.merge` or `R.mergeAll`.

**Comment by buzzdecafe** [comment]
I think this is a bit different, in order to support semigroups generally. For example, this works with `concat` today:

`concat(Tuple('a', 'c'), Tuple('b', 'd')) //=> Tuple('ab', 'cd')`

To capture the folding `concatAll` case, you need to know the empty type for the accumulator. In this example, that accumulator is `String * String`, so the implementation could be, e.g.

`const concatAll = reduce(concat, Tuple('', ''));`

... and so on, for any compliant semigroup. Since `concat` covers semigroups, I would expect its fold to do so as well.

**Comment by davidchambers** [comment]
Good point, @buzzdecafe. I was forgetting that in the `[]` case we need to know the empty value.

**Comment by CrossEye** [comment]
> Interestingly, if `R.concat` merged string maps (which it should) we wouldn't need `R.merge` or `R.mergeAll`.

Why string maps as opposed to other objects?  Although I hadn't thought about this before your note, ISTM that objects are generally monoidal with `{}` as the `empty` element, and our `merge` serving as `concat`.  Is there something special about string maps that would prompt us to treat them differently?

**Comment by davidchambers** [comment]
> Why string maps as opposed to other objects?

JavaScript objects are used in one of two ways: as string maps, or as records. It doesn't make sense to consider record types monoidal. Consider the following record type:

```javascript
//  type User = { id :: PositiveInteger, name :: String }

//  user :: User
var user = {id: 42, name: 'Alice'};

empty(user);
// => ???
```

If `empty(user)` evaluates to `{}` we've broken the [contract][1]:

> `empty` must return a value of the same Monoid

It could evaluate to `{id: 1, name: ''}` and be lawful, but `1` and `''` are arbitrary values of their respective types.


[1]: https://github.com/fantasyland/fantasy-land#monoid

**Comment by CrossEye** [comment]
This is the distinction between "Objects are monoidal" and "Records are monoidal".  It's clear that `User` above does not have an obvious monoid.  But ISTM that `Object` does, using `{}` for the identity and our current `merge` for the associative binary operator.

And of course Ramda does not have Sanctuary's type annotations, so we wouldn't have a way to know whether the user intends to use `{first: 'James', last: 'Bond'}` as a string map or a record, so how could you choose when to apply `concat` to it if you only wanted to handle the one case?  If you wanted to handle it for arbitrary objects, I don't see any issues, just _caveat emptor_ if you are extending a record type with some disallowed fields.

**Comment by davidchambers** [comment]
The implementation is likely to work for any old (or young!) objects. I'm just making a distinction between what we intend to support and what we would *actually* support.

**Comment by svozza** [comment]
> If you want this function it is trivial to implement as const concatAll = reduce(concat, []); which doesn't seem too onerous to me. 

This implementation will mean every intermediary collection will be cloned.

**Comment by buzzdecafe** [comment]
@svozza nice point -- ouch

**Comment by Bradcomp** [comment]
Given @buzzdecafe's comments about supporting semigroups generally, vs. just lists, doesn't this function already exist as `unnest`?

**Comment by paldepind** [comment]
I think @Bradcomp hit the nail on the head. Ramda already has `concatAll`. It's just called `unnest`. Maybe `unnest` is not the best name. But, I think it's fair to close this issue as the function requested in the OP is part of the library.

**Comment by GingerPlusPlus** [comment]
`concatAll` accepts Foldable containing Semigroups, whereas `R.unnest` accepts Chain containing Chains of the same type.

Thus, unlike `R.unnest`, `concatAll` behaves like `R.join('')` when passed array of strings.

It may be useful in [generic code](https://github.com/GingerPlusPlus/ramda-adjunct/commit/04a6be24e0d9482611ba615db5daeba60d8e75f8#diff-3d5f0b697b4314c9dbc340c8783967b5R28), that wants to work on both arrays and strings.

A `conctatAll` like that [is available in ramda-adjunct](https://char0n.github.io/ramda-adjunct/2.9.0/RA.html#.concatAll), discussion on adding it: https://github.com/char0n/ramda-adjunct/issues/109.

**Comment by zatziky** [comment]
As a workaround, one can use `[flatten()](https://ramdajs.com/docs/#flatten)` to concatenate multiple arrays

`flatten([[1,2], [2,3]])` → `[1,2,2,]`

**Comment by 3limin4t0r** [comment]
For those looking for a `concatAll` function that works with more than just arrays. You could use something like this:

``` JavaScript
const concatAll = R.converge(R.reduce(R.concat), [R.head, R.tail]);

concatAll([[1, 2], [3], [4, 5]]);        //=> [1, 2, 3, 4, 5]
concatAll(["Hello", " ", "World", "!"]); //=> "Hello World!"
concatAll(["Hello World!"]);             //=> "Hello World!"
concatAll([]);                           //=> undefined

// ramda-adjunct
concatAll([RA.Identity.of(3), RA.Identity.of(5), RA.Identity.of(4)]);
//=> Identity {value: 12}
concatAll([RA.Identity.of("a"), RA.Identity.of("b"), RA.Identity.of("c")]);
//=> Identity {value: "abc"}
```
